Native image fails to start: missing proxy registration for the OSHI NVML library

The native nightly (amd64) fails to boot. `DefaultGpuLoadMetrics` samples the GPU from its
constructor, OSHI reaches the NVML JNA binding, and native-image aborts because the JDK proxy
for `oshi.jna.common.Nvml$NvmlLibrary` was never registered:

```
Caused by: org.graalvm.nativeimage.MissingReflectionRegistrationError: Cannot reflectively access the proxy class inheriting ['oshi.jna.common.Nvml$NvmlLibrary'].
    at java.base@25.0.2/java.lang.reflect.Proxy.newProxyInstance(Proxy.java:924)
    at com.sun.jna.Native.load(Native.java:620)
    at oshi.util.gpu.NvmlUtilJNA$Holder.<clinit>(NvmlUtilJNA.java:64)
    at oshi.util.gpu.NvmlUtilJNA.isAvailable(NvmlUtilJNA.java:343)
    at oshi.hardware.platform.linux.LinuxGpuStatsJNA.nvmlIsAvailable(LinuxGpuStatsJNA.java:23)
    at oshi.hardware.common.platform.linux.LinuxGpuStats.findNvmlDevice(LinuxGpuStats.java:167)
    at oshi.hardware.common.platform.linux.LinuxGpuStats.getVramUsed(LinuxGpuStats.java:226)
    at com.krillsson.sysapi.core.metrics.defaultimpl.DefaultGpuLoadMetrics.sample(DefaultGpuLoadMetrics.kt:69)
    at com.krillsson.sysapi.core.metrics.defaultimpl.DefaultGpuLoadMetrics.<init>(DefaultGpuLoadMetrics.kt:40)
```

`NvmlUtilJNA$Holder` guards its `Native.load` against `UnsatisfiedLinkError`/`NoClassDefFoundError`,
so on the JVM a machine without NVML just reports "not available". `MissingReflectionRegistrationError`
is neither, so it escapes the static initialiser and takes the whole application context down.

Two things to fix:

1. Register the NVML JDK proxy (and the `Nvml` JNA structures, so the code path also works on a
   host that actually has NVML) in the native-image metadata.
2. Make `DefaultGpuLoadMetrics` fail-soft. GPU sampling should never be able to prevent the agent
   from starting.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,6 +1,7 @@
 ### Unreleased
 
 - Update OSHI to v7.4.3 (fixes GPU bugs on a Linux host)
+- Fix: the native image failed to start on any Linux host with a graphics card, because the NVIDIA NVML binding was missing its native-image proxy registration. GPU sampling can also no longer prevent the agent from starting
 
 ### 0.41.1
 

--- a/src/main/kotlin/com/krillsson/sysapi/RuntimeHints.kt
+++ b/src/main/kotlin/com/krillsson/sysapi/RuntimeHints.kt
@@ -4,6 +4,7 @@ import com.krillsson.sysapi.config.MdnsConfiguration
 import org.springframework.aot.hint.MemberCategory
 import org.springframework.aot.hint.RuntimeHints
 import org.springframework.aot.hint.RuntimeHintsRegistrar
+import oshi.jna.common.Nvml
 
 class RuntimeHint : RuntimeHintsRegistrar {
     override fun registerHints(hints: RuntimeHints, classLoader: ClassLoader?) {
@@ -14,5 +15,32 @@ class RuntimeHint : RuntimeHintsRegistrar {
                 MemberCategory.INVOKE_PUBLIC_METHODS,
                 MemberCategory.INVOKE_PUBLIC_CONSTRUCTORS
             )
+        registerNvmlHints(hints)
+    }
+
+    /**
+     * OSHI probes NVIDIA GPUs through a JNA binding, which builds a JDK proxy over
+     * [Nvml.NvmlLibrary]. OSHI guards the library load against `UnsatisfiedLinkError`, but an
+     * unregistered proxy fails with `MissingReflectionRegistrationError` instead, which escapes
+     * that guard and aborts startup on every Linux host that has a graphics card.
+     */
+    private fun registerNvmlHints(hints: RuntimeHints) {
+        hints.proxies().registerJdkProxy(Nvml.NvmlLibrary::class.java)
+        hints.reflection().registerType(
+            Nvml.NvmlLibrary::class.java,
+            MemberCategory.INVOKE_PUBLIC_METHODS
+        )
+        listOf(
+            Nvml.NvmlMemory::class.java,
+            Nvml.NvmlUtilization::class.java,
+            Nvml.NvmlPciInfo::class.java
+        ).forEach { structure ->
+            hints.reflection().registerType(
+                structure,
+                MemberCategory.DECLARED_FIELDS,
+                MemberCategory.INVOKE_DECLARED_CONSTRUCTORS,
+                MemberCategory.INVOKE_PUBLIC_METHODS
+            )
+        }
     }
 }

--- a/src/main/kotlin/com/krillsson/sysapi/core/metrics/defaultimpl/DefaultGpuLoadMetrics.kt
+++ b/src/main/kotlin/com/krillsson/sysapi/core/metrics/defaultimpl/DefaultGpuLoadMetrics.kt
@@ -28,6 +28,7 @@ class DefaultGpuLoadMetrics(private val hal: HardwareAbstractionLayer) {
     private val previousSampledAt = mutableMapOf<String, Long>()
     private val names = mutableMapOf<String, String>()
     private val vramTotals = mutableMapOf<String, Long>()
+    private val unreadableDevices = mutableSetOf<String>()
 
     @Volatile
     var gpuLoads: List<GpuLoad> = emptyList()
@@ -59,23 +60,42 @@ class DefaultGpuLoadMetrics(private val hal: HardwareAbstractionLayer) {
         return gpuLoads.firstOrNull { it.id == id }
     }
 
+    /**
+     * Never throws: a GPU we cannot read is reported as missing rather than being allowed to take
+     * down the metric that samples it — or, because the first sample happens in the constructor,
+     * the whole application context.
+     */
     private fun sample(): List<GpuLoad> {
-        refreshSessions()
-        return sessions.map { (deviceId, stats) ->
-            GpuLoad(
-                id = deviceId,
-                name = names[deviceId].orEmpty(),
-                coreLoad = utilization(deviceId, stats, System.nanoTime()),
-                vramUsedBytes = stats.vramUsed,
-                vramTotalBytes = vramTotals[deviceId] ?: -1L,
-                health = GpuHealth(
-                    temperature = stats.temperature,
-                    fanPercent = stats.fanSpeedPercent,
-                    powerDraw = stats.powerDraw,
-                    coreClockMhz = stats.coreClockMhz,
-                    memoryClockMhz = stats.memoryClockMhz
-                )
+        runCatching { refreshSessions() }
+            .onFailure { logger.warn("Unable to enumerate graphics cards", it) }
+        return sessions.mapNotNull { (deviceId, stats) ->
+            runCatching { read(deviceId, stats) }
+                .onSuccess { unreadableDevices.remove(deviceId) }
+                .onFailure { reportUnreadable(deviceId, it) }
+                .getOrNull()
+        }
+    }
+
+    private fun read(deviceId: String, stats: GpuStats): GpuLoad {
+        return GpuLoad(
+            id = deviceId,
+            name = names[deviceId].orEmpty(),
+            coreLoad = utilization(deviceId, stats, System.nanoTime()),
+            vramUsedBytes = stats.vramUsed,
+            vramTotalBytes = vramTotals[deviceId] ?: -1L,
+            health = GpuHealth(
+                temperature = stats.temperature,
+                fanPercent = stats.fanSpeedPercent,
+                powerDraw = stats.powerDraw,
+                coreClockMhz = stats.coreClockMhz,
+                memoryClockMhz = stats.memoryClockMhz
             )
+        )
+    }
+
+    private fun reportUnreadable(deviceId: String, throwable: Throwable) {
+        if (unreadableDevices.add(deviceId)) {
+            logger.warn("Unable to read load for GPU {}, skipping it until it recovers", deviceId, throwable)
         }
     }
 
@@ -117,6 +137,8 @@ class DefaultGpuLoadMetrics(private val hal: HardwareAbstractionLayer) {
             runCatching { sessions[deviceId]?.close() }
             sessions.remove(deviceId)
             previousTicks.remove(deviceId)
+            previousSampledAt.remove(deviceId)
+            unreadableDevices.remove(deviceId)
         }
     }
 

